syncutil/synccache/synccache.go

```diff
@@ -0,0 +1,97 @@
+package synccache
+
+import (
+	"context"
+	"sync"
+
+	"github.com/earthly/earthly/syncutil/metacontext"
+	"github.com/pkg/errors"
+)
+
+// Constructor is a func that is used to construct a cache value, given a key.
+type Constructor func(ctx context.Context, key interface{}) (value interface{}, err error)
+
+type entry struct {
+	metaCtx *metacontext.MetaContext
+
+	constructed chan struct{}
+	err         error
+	value       interface{}
+}
+
+// SyncCache is an object which can be used to create singletons stored in a key-value store.
+type SyncCache struct {
+	mu    sync.Mutex
+	store map[interface{}]*entry
+}
+
+// New creates an empty SyncCache.
+func New() *SyncCache {
+	return &SyncCache{
+		store: make(map[interface{}]*entry),
+	}
+}
+
+// Do executes the constructor, if a value for key hasn't already been constructed.
+func (sc *SyncCache) Do(ctx context.Context, key interface{}, c Constructor) (interface{}, error) {
+	e, found := sc.getEntry(ctx, key)
+	if !found {
+		// We need to construct this.
+		go func() {
+			// The metaCtx will ensure that this stays alive even if the original Do has
+			// been canceled, thanks to the metaCtx. This is canceled only when ALL of
+			// the Do's are canceled.
+			e.value, e.err = c(e.metaCtx, key)
+			// Don't cache context canceled. Whoever is currently waiting will still get this,
```

Comment:
This is for a case that doesn't apply to us but is in there for correctness. The thinking is that if *some* of the work has been canceled, but not all, it's possible that the `SyncCache` entry to not finish processing as a result. However, if some other part has not been canceled and ends up building the same entry, it shouldn't forever say "I'm canceled therefore you can never construct me again" if it comes with a perfectly valid context (not canceled).

An example of such a situation is if the work were part of two completely distinct network requests (again, not our situation specifically), of which only one of them was canceled, and the second one comes along after some time.

When using `Ctrl+C` it's not possible for things to not propagate through all ctx's because it cancels the very root, which affects every context we've ever created in that session.

Reply:
Awesome. Thanks for clarifying!